### PR TITLE
Features/superadmin wants to fill out profile

### PR DESCRIPTION
**Pull request by casaper**
This PR needs to be rebased after https://github.com/panter/aoz-003/pull/5 because it will conflict with it otherwise

### CODE REVIEW

File: app/controllers/application_controller.rb [hunk 1]
```diff
@@ -3,5 +3,13 @@ class ApplicationController < ActionController::Base
 
   before_action :authenticate_user!
 
+  def after_sign_in_path_for(current_user)
+    if current_user.profile.blank?
+      new_profile_path
+    else
+      current_user
```
Comment (kitsane) [inline app/controllers/application_controller.rb]: Set `root`in routes to `users_path` and redirect to `root_path`in the else clause

File: app/controllers/application_controller.rb [hunk 2]
```diff
@@ -3,5 +3,13 @@ class ApplicationController < ActionController::Base
...
+    end
+  end
+
   def logged_in; end
```
Comment (kitsane) [inline app/controllers/application_controller.rb]: can be deleted, also the view file in `views/applications`

File: app/models/user.rb [hunk 3]
```diff
@@ -2,6 +2,8 @@ class User < ApplicationRecord
   devise :database_authenticatable, :recoverable, :rememberable,
     :trackable, :validatable
 
+  has_one :profile
```
Comment (kitsane) [inline app/models/user.rb]: I think, just to be shure, add `dependent: :destroy`

File: app/views/users/_form.html.slim [hunk 4]
```diff
@@ -5,8 +5,12 @@
         = f.error_notification
   .row
     .col-xs-12
-      = f.input :email
-      = f.input :role, collection: [User::SUPERADMIN], input_html: { class: 'dropdown' }
+      - if current_user == user
+        = f.input :email
```
Comment (kitsane) [inline app/views/users/_form.html.slim]: Sorry, i moved it to far in, email should be before the if

File: config/routes.rb [hunk 5]
```diff
@@ -2,5 +2,7 @@
   devise_for :users
   resources :users, except: :destroy
 
+  resources :profiles, except: %i[destroy index]
```
Comment (kitsane) [inline config/routes.rb]: Use `[:destroy, :index]`

File: config/routes.rb [hunk 6]
```diff
@@ -2,5 +2,7 @@
...
+
   root 'application#logged_in'
```
Comment (kitsane) [inline config/routes.rb]: Change to `users#index`

File: db/migrate/20170411150256_create_profiles.rb [hunk 7]
```diff
@@ -0,0 +1,20 @@
+class CreateProfiles < ActiveRecord::Migration[5.1]
+  def change
+    create_table :profiles do |t|
+      t.references :user, foreign_key: true
+      t.string :first_name
+      t.string :last_name
+      t.string :phone
+      t.string :picture
```
Comment (kitsane) [inline db/migrate/20170411150256_create_profiles.rb]: Use paperclip gem for attachments

File: test/controllers/profiles_controller_test.rb [hunk 8]
```diff
@@ -0,0 +1,27 @@
+require 'test_helper'
```
Comment (kitsane) [inline test/controllers/profiles_controller_test.rb]: Drop this tests

File: test/system/sessions_test.rb [hunk 9]
```diff
@@ -2,7 +2,8 @@
 
 class SessionsTest < ApplicationSystemTestCase
   setup do
-    create :user
+    profile = create :profile
+    @superadmin = User.find profile.user_id
```
Comment (kitsane) [inline test/system/sessions_test.rb]: `@superadmin = profile.user`

File: test/factories/users.rb [hunk 10]
```diff
@@ -4,4 +4,10 @@
     password 'asdfasdf'
     role 'superadmin'
   end
+
+  factory :user_noprofile, class: User do
```
Comment (kitsane) [inline test/factories/users.rb]: I think it's better to have just a user factory and a user factory with_profile. You can use [Factory Girl Callbacks](https://github.com/thoughtbot/factory_girl/blob/master/GETTING_STARTED.md#callbacks)

File: app/controllers/application_controller.rb (hunk 1 above)
Comment (casaper) [inline app/controllers/application_controller.rb]: done

File: app/controllers/application_controller.rb (hunk 2 above)
Comment (casaper) [inline app/controllers/application_controller.rb]: - [x]

File: app/views/users/_form.html.slim (hunk 4 above)
Comment (casaper) [inline app/views/users/_form.html.slim]: So nothing to do for me here?

File: config/routes.rb (hunk 5 above)
Comment (casaper) [inline config/routes.rb]: - [x] done

File: config/routes.rb (hunk 6 above)
Comment (casaper) [inline config/routes.rb]: - [x] done

File: db/migrate/20170411150256_create_profiles.rb (hunk 7 above)
Comment (casaper) [inline db/migrate/20170411150256_create_profiles.rb]: - [x] done

File: test/controllers/profiles_controller_test.rb (hunk 8 above)
Comment (casaper) [inline test/controllers/profiles_controller_test.rb]: - [x] done

**Comment by casaper** [comment]
- [ ] Merge Request on github
- [ ] Story is under test
- [ ] Acceptance criterias are met
- [ ] Code is reviewed
- [ ] Code quality checks are green
- [ ] Ci is green
- [ ] Mobile view is usable
- [ ] Functionality is tested in Target Browsers
- [ ] Story is on done in pm tool
- [ ] Readme updated if needed
- [ ] Styled with Bootstrap v3
- [ ] Seeds are available
- [ ] German translations added

File: test/system/profiles_test.rb [hunk 11]
```diff
@@ -0,0 +1,83 @@
+require 'application_system_test_case'
+
+class ProfilesTest < ApplicationSystemTestCase
+  def setup
+    @superadmin = create :user_with_profile
```
Comment (kitsane) [inline test/system/profiles_test.rb]: @user_with_profile

File: test/system/profiles_test.rb [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+    @noprofile = create :user
```
Comment (kitsane) [inline test/system/profiles_test.rb]: @user_without_profile

File: test/system/profiles_test.rb [hunk 13]
```diff
@@ -0,0 +1,83 @@
...
+  end
+
+  test 'when first login displays profile form' do
+    visit new_user_session_path
+    fill_in 'Email', with: @noprofile.email
+    fill_in 'Password', with: 'asdfasdf'
+    click_on 'Log in'
```
Comment (kitsane) [inline test/system/profiles_test.rb]: When you know what it is use the more specific method to click, in this case `click_link`but good to know, that there is something to click whatever, cool

File: test/system/sessions_test.rb [hunk 14]
```diff
@@ -2,7 +2,7 @@
 
 class SessionsTest < ApplicationSystemTestCase
   setup do
-    create :user
+    @superadmin = create :user_with_profile
```
Comment (kitsane) [inline test/system/sessions_test.rb]: This is a session test without profile stuff, so we don't depend on the user to be a superadmin, just use `@user = create :user`

File: app/assets/stylesheets/_main.scss [hunk 15]
```diff
@@ -1,6 +1,18 @@
 @import 'bootstrap_overwrites';
 @import 'bootstrap/variables';
 
+html, body {
+  display: flex;
+  flex-direction: column;
+  min-height: 100%;
+  height: 100vh;
```
Comment (kitsane) [inline app/assets/stylesheets/_main.scss]: `min-height` is allready 100% is the `height` still needed?

File: app/models/profile.rb [hunk 16]
```diff
@@ -0,0 +1,10 @@
+class Profile < ApplicationRecord
+  belongs_to :user
+  has_attached_file :avatar, styles: { thumb: '100x100#' }
+
+  validates_attachment :avatar, content_type: {
+    content_type: ['image/jpeg', 'image/gif', 'image/png']
+  }
+  validates_attachment_file_name :avatar, matches: [/png\z/, /jpe?g\z/]
```
Comment (kitsane) [inline app/models/profile.rb]: `validates_attachment`and `validates_attachment_file_name`is old validations syntax style, use:

```ruby
validates_attachment :avatar, presence: true, content_type: /\Aimage/
```
The filename validation is not needed as it is handeled through `content_type`

File: app/views/profiles/_form.html.slim [hunk 17]
```diff
@@ -0,0 +1,24 @@
...
+    .row
+      .col-xs-12
+        = f.error_notification
+  .row
+    .col-xs-12
+      = f.input :user_id, as: :hidden
+      = f.input :first_name
+      = f.input :last_name
+      = f.input :phone
+      = f.input :avatar, class: 'form-control file avatar optional'
```
Comment (kitsane) [inline app/views/profiles/_form.html.slim]: I think this is cleaner, `= f.input :avatar, as: :file, class: 'form-control'`

File: app/views/profiles/show.html.slim [hunk 18]
```diff
@@ -0,0 +1,36 @@
+h1 My Profile
+
+h2 Information
```
Comment (kitsane) [inline app/views/profiles/show.html.slim]: Delete this title

File: app/views/users/_form.html.slim [hunk 19]
```diff
@@ -5,8 +5,12 @@
...
+        = f.input :password
+      - if current_user.role == User::SUPERADMIN
```
Comment (kitsane) [inline app/views/users/_form.html.slim]: I think you did a helper method like `current_user.superadmin?`

File: test/application_system_test_case.rb [hunk 20]
```diff
@@ -1,12 +1,27 @@
...
+    window_size: DIMENSIONS
+  )
+end
+
+Capybara.register_driver :poltergeist do |app|
+  Capybara::Poltergeist::Driver.new(
+    app,
+    screen_size: DIMENSIONS,
+    window_size: DIMENSIONS
+  )
```
Comment (kitsane) [inline test/application_system_test_case.rb]: Is this still needed or can we get rid of it?

File: test/system/sessions_test.rb [hunk 21]
```diff
@@ -17,23 +17,27 @@ class SessionsTest < ApplicationSystemTestCase
...
+    fill_in 'Email', with: @superadmin.email
     fill_in 'Password', with: 'asdfasdf'
-    click_button 'Log in'
-
+    click_on 'Log in'
     assert page.has_current_path? root_path
     assert page.has_text? 'Signed in successfully.'
-    assert page.has_link? 'superadmin@example.com'
+
+    click_on 'Toggle navigation' if page.has_button? 'Toggle navigation'
```
Comment (kitsane) [inline test/system/sessions_test.rb]: We should get rid of this if, we control the dimensions of the window so we never should have the need to use something like this
Tests in general should be as stupid as possible

File: test/system/sessions_test.rb [hunk 22]
```diff
@@ -17,23 +17,27 @@ class SessionsTest < ApplicationSystemTestCase
...
 
   test 'sign out current user' do
-    user = User.first
-    login_as user, scope: :user
+    login_as @superadmin, scope: :user
     visit root_path
 
-    click_link 'superadmin@example.com'
+    # trigger hamburger on mobile views
+    click_on 'Toggle navigation' if page.has_button? 'Toggle navigation'
```
Comment (kitsane) [inline test/system/sessions_test.rb]: Same as above

**Pull request merged**